src/main/java/be/seeseemelk/mockbukkit/MockUnsafeValues.java

```diff
@@ -296,32 +301,119 @@ public Entity deserializeEntity(byte[] data, World world, boolean preserveUUID)
+	 * @param type the entity to translate
+	 * @return a string of the structure "item.{@literal <namespace>}.{@literal <entity_type>}" (e.g. "entity.minecraft.pig").
+	 */
 	@Override
 	public String getTranslationKey(EntityType type)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		Preconditions.checkArgument(type.getName() != null, "Invalid name of EntityType %s for translation key", type);
+		Arrays.stream(EntityType.values())
+				.filter(entityType ->type.getName().equals(entityType.getName()))
+				.findFirst()
+				.orElseThrow();
+		return formatTranslatable("entity", type);
 	}
 
+	/**
+	 * Gets the translation key for a {@link ItemStack} that {@link Material#isItem() is an item}.
+	 * @param itemStack the itemstack to translate
+	 * @return a string of the structure "item.{@literal <namespace>}.{@literal <material>}" (e.g. "item.minecraft.carrot"), or null if not an item.
+	 */
 	@Override
 	public String getTranslationKey(ItemStack itemStack)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		if(itemStack.getType().isItem()) {
+			Material material = itemStack.getType();
+			if(!material.isItem()) {
+				return null;
+			}
+			String edgeCaseHandledTranslationKey = handleTranslateItemEdgeCases(material);
+			if(edgeCaseHandledTranslationKey != null) {
+				return edgeCaseHandledTranslationKey;
+			}
+			return formatTranslatable("item", material, true);
+		}
+		else if(itemStack.getType().isBlock()) {
+			return getBlockTranslationKey(itemStack.getType());
+		}
+		else {
+			return null;
+		}
+	}
+
+	private String handleTranslateItemEdgeCases(Material material)
+	{
+		// edge cases: WHEAT and NETHER_WART are blocks, but still use the "item" prefix (therefore this check has to be done BEFORE the isBlock check below)
+		if(material == Material.WHEAT || material == Material.NETHER_WART) {
+			return formatTranslatable("item", material);
+		}
+		// edge case: If a translation key from an item is requested from anything that is also a block, the block translation key is always returned
+		// e.g: Material#STONE is a block (but also an obtainable item in the inventory). However, the translation key is always "block.minecraft.stone".
+		if(material.isBlock()) {
+			return formatTranslatable("block", material);
+		}
+		// not an edge case
+		return null;
+	}
+
+	private <T extends Keyed & Translatable> String formatTranslatable(String prefix, T translatable, boolean fromItemStack)
+	{
+		// enforcing Translatable is not necessary, but translating only makes sense when the object is really translatable by design.
+		String value = translatable.key().value();
+		if(translatable instanceof Material material) {
+			if(Tag.WALL_HANGING_SIGNS.isTagged(material) || Tag.WALL_SIGNS.isTagged(material) || value.endsWith("wall_banner") || value.endsWith("wall_torch") || value.endsWith("wall_skull") || value.endsWith("wall_head")) {
```

Comment:
Require help here. The tag registry returns null for `WALL_HANGING_SIGNS`. I checked and `WALL_HANGING_SIGNS` is missing in the json file for the tags.

Reply:
It's not currently implemented, I have been working on something that scans all tags in a real server and saves them to JSON files, but I haven't been able to get it to fully work yet. You simply have to ignore this for now. Accept a good enough solution for now.